Xamarin.Forms.Core/Shell/Shell.cs

```diff
@@ -1653,6 +1652,79 @@ bool ProposeNavigation(ShellNavigationSource source, ShellNavigationState propos
 			return !navArgs.Cancelled && navArgs.DeferralCount == 0;
 		}
 
+		ShellNavigationSource CalculateNavigationSource(ShellNavigationState current, NavigationRequest request)
+		{
+			if (request.StackRequest == NavigationRequest.WhatToDoWithTheStack.PushToIt)
+				return ShellNavigationSource.Push;
+
+			if (current == null)
+				return ShellNavigationSource.Unknown;
+
+			var targetUri = ShellUriHandler.ConvertToStandardFormat(this, request.Request.FullUri);
+			var currentUri = ShellUriHandler.ConvertToStandardFormat(this, current.FullLocation);
+			
+			var targetPaths = ShellUriHandler.RetrievePaths(targetUri.PathAndQuery);
+			var currentPaths = ShellUriHandler.RetrievePaths(currentUri.PathAndQuery);
+
+			var targetPathsLength = targetPaths.Length;
+			var currentPathsLength = currentPaths.Length;
+
+			if (targetPathsLength < 4 || currentPathsLength < 4)
+				return ShellNavigationSource.Unknown;
+
+			if (targetPaths[1] != currentPaths[1])
+				return ShellNavigationSource.ShellItemChanged;
+			if (targetPaths[2] != currentPaths[2])
+				return ShellNavigationSource.ShellSectionChanged;
+			if (targetPaths[3] != currentPaths[3])
+				return ShellNavigationSource.ShellContentChanged;
+
+			if (targetPaths.Length == currentPaths.Length)
+				return ShellNavigationSource.Unknown;
+
+			if(targetPaths.Length < currentPaths.Length)
+			{
+				for (var i = 0; i < targetPaths.Length; i++)
+				{
+					var targetPath = targetPaths[i];
+					if (targetPath != currentPaths[i])
+						break;
+
+					if (i == targetPaths.Length - 1)
+					{
+						if (targetPaths.Length == 4)
+							return ShellNavigationSource.PopToRoot;
+
+						return ShellNavigationSource.Pop;
+					}
+				}
+
+				if (targetPaths[targetPathsLength - 1] == currentPaths[currentPathsLength - 1])
+					return ShellNavigationSource.Remove;
+
+				if (targetPaths.Length == 4)
+					return ShellNavigationSource.PopToRoot;
+
+				return ShellNavigationSource.Pop;
+			}
+			else if(targetPaths.Length > currentPaths.Length)
+			{
+				for (var i = 0; i < currentPaths.Length; i++)
+				{
+					if (targetPaths[i] != currentPaths[i])
+						break;
+
+					if (i == targetPaths.Length - 1)
+						return ShellNavigationSource.Push;
+				}
+			}
+
+			if (targetPaths[targetPathsLength - 1] == currentPaths[currentPathsLength - 1])
+				return ShellNavigationSource.Insert;
+
+			return ShellNavigationSource.Push;
```

Comment:
@PureWeen  just a micro-opt avoiding to call the `get` method inside the props. (:
```suggestion
		var targetPathsLength = targetPaths.Length;
			var currentPathsLength = currentPaths.Length;

			if (targetPathsLength < 4 || currentPathsLength < 4)
				return ShellNavigationSource.Unknown;

			if (targetPaths[1] != currentPaths[1])
				return ShellNavigationSource.ShellItemChanged;
			if (targetPaths[2] != currentPaths[2])
				return ShellNavigationSource.ShellSectionChanged;
			if (targetPaths[3] != currentPaths[3])
				return ShellNavigationSource.ShellContentChanged;
'
			if (targetPathsLength == currentPathsLength)
				return ShellNavigationSource.Unknown;

			if(targetPathsLength < currentPathsLength)
			{
				for (var i = 0; i < targetPathsLength; i++)
				{
					var targetPath = targetPaths[i];
					if (targetPath != currentPaths[i])
						break;

					if (i == targetPathsLength - 1)
					{
						if (targetPathsLength == 4)
							return ShellNavigationSource.PopToRoot;

						return ShellNavigationSource.Pop;
					}
				}

				if (targetPaths[targetPathsLength - 1] == currentPaths[currentPathsLength - 1])
					return ShellNavigationSource.Remove;

				if (targetPathsLength == 4)
					return ShellNavigationSource.PopToRoot;

				return ShellNavigationSource.Pop;
			}
			else if(targetPathsLength > currentPathsLength)
			{
				for (var i = 0; i < currentPathsLength; i++)
				{
					if (targetPaths[i] != currentPaths[i])
						break;

					if (i == targetPathsLength - 1)
						return ShellNavigationSource.Push;
				}
			}

			if (targetPaths[targetPathsLength - 1] == currentPaths[currentPathsLength - 1])
				return ShellNavigationSource.Insert;

			return ShellNavigationSource.Push;
```

Reply:
👍